### DISCUSSION TITLE
Does Dexie support any in-memory data storage or caching?

### DISCUSSION THREAD

**Discussion by tomkennedy22** [dexie/Dexie.js#1650] [Q&A] [answered]
Hey all, I've been using Dexie for quite some time and really enjoy the product. 
In my web app using Dexie, I jump around many functions and read & write to the same Dexie tables quite often. These functions are discrete and run asynchronously. 
In my use, some tables have 10k+ rows, so reading & writing often can take some time. I was wondering if there's any functionality, either now or planned, that may assist with this. Being able to store data in-memory and query from there using the same Dexie syntax would be cool, and faster. And then there could be a 'write' trigger to sync from in-memory to disk. 

I know there are some packages out there right now that do similar things, but I do like Dexie a bit, so am not motivated to move.

Thanks!

**Comment by dfahlander** [comment, marked as answer]
Hi! Thanks for sharing your thoughts!

Dexie 4 is planned for start using caching for non-transactional readonly queries. I've written a few words about it on the [road map](https://dexie.org/roadmap/dexie4.0#cache) page. Having dexie being backed by in-memory data only is also a valid use case for testing etc. Today this can be accomplished using [fake-indexeddb](https://www.npmjs.com/package/fake-indexeddb) but it would probably be nice if it was built-in, or if there would be possible to do an addon for it, such a a memory-backed [DBCore](https://dexie.org/docs/DBCore/DBCore) (would be possible to implement already, but it's not possible to replace the database opening / upgrading parts yet).

**Reply by tomkennedy22** [reply]
Thanks @dfahlander, makes sense. I was looking around and found [LokiJS](https://github.com/techfort/LokiJS) to cover some caching capabilities, although I'm not sure that it is maintained any longer. 

Looking forward to Dexie 4! Thanks for all of your work!
